src/bindings/python/flux/hostlist.py

```diff
@@ -0,0 +1,208 @@
+                ffi,
+                lib,
+                handle=handle,
+                match=ffi.typeof("struct hostlist *"),
+                prefixes=["hostlist_"],
+                destructor=lib.hostlist_destroy,
+            )
+
+    def __init__(self, arg="", handle=None):
+        """
+        Create a new Hostlist object from RFC 29 Hostlist string or an
+        Iterable containing a set of RFC 29 Hostlist strings:
+
+        :param arg: A string or Iterable containing one or more hosts
+                    encoded in RFC 29 hostlist format
+
+        E.g.
+
+        >>> hl = Hostlist()
+        >>> hl = Hostlist("host")
+        >>> hl = Hostlist("host[0-10]")
+        >>> hl = Hostlist([ "foo1", "foo2" ])
+
+        """
+        #  If no `struct hostlist *` handle passed in, then create
+        #   a new handle from string argument
+        if handle is None:
+            if isinstance(arg, str):
+                handle = lib.hostlist_decode(arg.encode("utf-8"))
+                if handle == ffi.NULL:
+                    raise ValueError(f"Invalid hostlist: '{arg}'")
+            elif isinstance(arg, collections.Iterable):
+                handle = lib.hostlist_create()
+                for hosts in arg:
+                    if not isinstance(hosts, str):
+                        typestr = type(hosts)
+                        raise TypeError(
+                            f"Hostlist(): expected string or Iterable, got {typestr}"
+                        )
+                    result = lib.hostlist_append(handle, hosts.encode("utf-8"))
+                    if result < 0:
+                        raise ValueError(f"Invalid hostlist: '{hosts}'")
+            else:
+                typestr = type(arg)
+                raise TypeError(
+                    f"Hostlist(): expected string or Iterable, got {typestr}"
+                )
+        super().__init__()
+        self.pimpl = self.InnerWrapper(handle)
+
+    def __str__(self):
+        return self.encode()
+
+    def __repr__(self):
+        return f"Hostlist('{self}')"
+
+    def __len__(self):
+        return self.pimpl.count()
+
+    def __getitem__(self, index):
+        """Index and slice a hostlist
+
+        Works like normal Python list indexing, including slices:
+
+        >>> hl = Hostlist("foo[0-9]")
+        >>> hl[0]
+        'foo0'
+        >>> hl[9]
+        'foo9'
+        >>> hl[-1]
+        'foo9'
+        >>> hl[8:]
+        ['foo8', 'foo9']
+        >>> hl[1:3]
+        ['foo1', 'foo2']
+
+        """
+        if isinstance(index, numbers.Integral):
+            if index < 0:
+                index = len(self) + index
+            if 0 <= index < len(self):
+                # N.B. wrapper class already calls ffi.string() on result:
+                return self.pimpl.nth(index).decode("utf-8")
+            raise IndexError("Hostlist index out of range")
+
+        if isinstance(index, slice):
+            return [self[n] for n in range(len(self))[index]]
```

Comment:
Minor nitpick:
```suggestion
            return [self[n] for n in range(index.start, index.stop, index.step)]
```
will avoid constructing unnecessary elements if the slice starts at index > 0.

Sidenote: it's kind of annoying that `range` doesn't support taking a slice as an argument.  And that slices don't support `*` expansion.

Reply:
@SteVwonder, I did try this approach, but it isn't so simple.

For a slice, I found that any of `start`, `stop`, or `step` can be `None` (or maybe only two of them, I don't remember), and `range` does not support `None` for some arguments. It ended being a bit of mess trying to handle all the corner cases, and `range(len(self))[index]` seemed to be a lot cleaner.

However, your comment made think, why don't we just make a function in `hostlist.py` that will operate like `range()`, but _does_ support slices? I'll give that a try.

Reply:
itertools.islice?

```python
[i for i in it.islice(range(10), s.start, s.stop, s.step)]
```

Reply:
Thanks! It would have taken me awhile to find that one.

Reply:
Hm `islice` doesn't support negative values so this won't work:
```python
>>> hl = Hostlist("fluke[0-7]")
>>> hl[::-1]
['fluke7', 'fluke6', 'fluke5', 'fluke4', 'fluke3', 'fluke2', 'fluke1', 'fluke0']
```

with `islice`:
```python
>>> hl = Hostlist("fluke[0-7]")
>>> hl[::-1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/g/g0/grondo/git/flux-core.git/src/bindings/python/flux/hostlist.py", line 134, in __getitem__
    return [i for i in islice(self, index.start, index.stop, index.step)]
ValueError: Step for islice() must be a positive integer or None.

```

Reply:
Wow, dang.  I thought that was supposed to give you the full slice setup. 😞

Reply:
Is it just the step that will not take negatives? If so, one last crack:

```python
it = self
if s.step < 0:
  it = reversed(self)
[i for i in it.islice(it, s.start, s.stop, abs(s.step))]
```

Now mainly doing this because this feels like it shouldn't be hard, and yet... and it's bugging me. 😆

Reply:
This simple wrapper for `range()` seems to work:
```python
def rangeslice(index, maxi):
    step = index.step
    if step is None:
        step = 1
    start = index.start
    if start is None:
        start = 0 if step > 0 else maxi
    stop = index.stop
    if stop is None:
        stop = maxi + 1 if step > 0 else -1
    return range(start, stop, step)
```

Reply:
Oh sorry @trws, Github didn't refresh the page so I missed your suggestion above. I'll give that a try.

Reply:
Your solution works perfectly nicely as far as I'm concerned, it feels so odd to need to work around this.

Reply:
> Is it just the step that will not take negatives? If so, one last crack:

None of the values can take negatives according to the  [docs](https://docs.python.org/3.8/library/itertools.html#itertools.islice):

> Unlike regular slicing, islice() does not support negative values for start, stop, or step.